crates/egui/src/widgets/image.rs

```diff
@@ -796,57 +798,90 @@ pub fn paint_texture_at(
         .ok_or("Failed to find index separator '#'")?;
-    let index: usize = index
-        .parse()
-        .map_err(|_err| format!("Failed to parse gif frame index: {index:?} is not an integer"))?;
+    let index: usize = index.parse().map_err(|_err| {
+        format!("Failed to parse animated image frame index: {index:?} is not an integer")
+    })?;
     Ok((uri, index))
 }
 
-/// checks if uri is a gif file
-fn is_gif_uri(uri: &str) -> bool {
-    uri.ends_with(".gif") || uri.contains(".gif#")
-}
-
-/// checks if bytes are gifs
-pub fn has_gif_magic_header(bytes: &[u8]) -> bool {
-    bytes.starts_with(b"GIF87a") || bytes.starts_with(b"GIF89a")
-}
+/// Calculates at which frame the animated image is
+fn animated_image_frame_index(ctx: &Context, uri: &str) -> usize {
+    let now = ctx.input(|input| Duration::from_secs_f64(input.time));
 
-/// calculates at which frame the gif is
-fn gif_frame_index(ctx: &Context, uri: &str) -> usize {
-    let now = ctx.input(|i| Duration::from_secs_f64(i.time));
+    let durations: Option<FrameDurations> = ctx.data(|data| data.get_temp(Id::new(uri)));
 
-    let durations: Option<GifFrameDurations> = ctx.data(|data| data.get_temp(Id::new(uri)));
     if let Some(durations) = durations {
-        let frames: Duration = durations.0.iter().sum();
+        let frames: Duration = durations.all().sum();
         let pos_ms = now.as_millis() % frames.as_millis().max(1);
+
         let mut cumulative_ms = 0;
-        for (i, duration) in durations.0.iter().enumerate() {
+
+        for (index, duration) in durations.all().enumerate() {
             cumulative_ms += duration.as_millis();
+
             if pos_ms < cumulative_ms {
                 let ms_until_next_frame = cumulative_ms - pos_ms;
                 ctx.request_repaint_after(Duration::from_millis(ms_until_next_frame as u64));
-                return i;
+                return index;
             }
         }
+
         0
     } else {
         0
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
-/// Stores the durations between each frame of a gif
-pub struct GifFrameDurations(pub Arc<Vec<Duration>>);
+/// Checks if uri is a gif file
+fn is_gif_uri(uri: &str) -> bool {
+    uri.ends_with(".gif") || uri.contains(".gif#")
+}
+
+/// Checks if bytes are gifs
+pub fn has_gif_magic_header(bytes: &[u8]) -> bool {
+    bytes.starts_with(b"GIF87a") || bytes.starts_with(b"GIF89a")
+}
+
+/// Checks if uri is a webp file
+fn is_webp_uri(uri: &str) -> bool {
+    uri.ends_with(".webp") || uri.contains(".webp#")
+}
+
+/// Checks if bytes are webp
+pub fn has_webp_header(bytes: &[u8]) -> bool {
+    bytes.len() >= 12 && &bytes[0..4] == b"RIFF" && &bytes[8..12] == b"WEBP"
+}
+
+fn is_animated_image_uri(uri: &str) -> bool {
+    is_gif_uri(uri) || is_webp_uri(uri)
+}
+
+fn are_animated_image_bytes(uri: &str, bytes: &[u8]) -> bool {
+    (is_gif_uri(uri) && has_gif_magic_header(bytes)) || (is_webp_uri(uri) && has_webp_header(bytes))
```

Comment:
I think we should only check the bytes here; the uri may or may not contain the file ending, but if the magic bytes are correct it should always be fine

Reply:
I'm fine with that